pkg/ipam/allocator/operator/operator.go

```diff
@@ -0,0 +1,116 @@
+// limitations under the License.
+
+package operator
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	operatorMetrics "github.com/cilium/cilium/operator/metrics"
+	operatorOption "github.com/cilium/cilium/operator/option"
+	ipPkg "github.com/cilium/cilium/pkg/ip"
+	"github.com/cilium/cilium/pkg/ipam"
+	"github.com/cilium/cilium/pkg/ipam/allocator"
+	ipamMetrics "github.com/cilium/cilium/pkg/ipam/metrics"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/trigger"
+
+	"github.com/cilium/ipam/cidrset"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ipam-allocator-operator")
+
+// AllocatorOperator is an implementation of IPAM allocator interface for Cilium
+// IPAM.
+type AllocatorOperator struct {
+	v4CIDRSet, v6CIDRSet []ipam.CIDRAllocator
+}
+
+// Init sets up Cilium allocator based on given options
+func (a *AllocatorOperator) Init() error {
+	if len(operatorOption.Config.IPAMOperatorV4CIDR) != 0 {
+		if !option.Config.EnableIPv4 {
+			return errors.New("IPv4CIDR can not be set if IPv4 is not enabled")
+		}
+		v4Allocators, err := newCIDRSets(false, operatorOption.Config.IPAMOperatorV4CIDR, operatorOption.Config.NodeCIDRMaskSizeIPv4)
+		if err != nil {
+			return fmt.Errorf("unable to initialize IPv4 allocator %w", err)
+		}
+		a.v4CIDRSet = v4Allocators
+	}
+	if len(operatorOption.Config.IPAMOperatorV6CIDR) != 0 {
+		if !option.Config.EnableIPv6 {
+			return errors.New("IPv6CIDR can not be set if IPv6 is not enabled")
+		}
+		v6Allocators, err := newCIDRSets(true, operatorOption.Config.IPAMOperatorV6CIDR, operatorOption.Config.NodeCIDRMaskSizeIPv6)
+		if err != nil {
+			return fmt.Errorf("unable to initialize IPv6 allocator %w", err)
+		}
+		a.v6CIDRSet = v6Allocators
+	}
+	if len(a.v4CIDRSet)+len(a.v6CIDRSet) == 0 {
+		return fmt.Errorf("either '%s' or '%s' need to be set", operatorOption.IPAMOperatorV4CIDR, operatorOption.IPAMOperatorV6CIDR)
+	}
+	return nil
+}
+
+// Start kicks of Operator allocation.
+func (a *AllocatorOperator) Start(updater ipam.CiliumNodeGetterUpdater) (allocator.NodeEventHandler, error) {
+	log.Info("Starting Operator IP allocator...")
+
+	var (
+		iMetrics trigger.MetricsObserver
+	)
+
+	if operatorOption.Config.EnableMetrics {
+		iMetrics = ipamMetrics.NewTriggerMetrics(operatorMetrics.Namespace, "k8s_sync")
+	} else {
+		iMetrics = &ipamMetrics.NoOpMetricsObserver{}
+	}
+
+	nodeManager := ipam.NewNodesPodCIDRManager(a.v4CIDRSet, a.v6CIDRSet, updater, iMetrics)
+
+	return nodeManager, nil
+}
+
+func newCIDRSets(isV6 bool, strCIDRs []string, maskSize int) ([]ipam.CIDRAllocator, error) {
+	cidrAllocators := make([]ipam.CIDRAllocator, 0, len(strCIDRs))
+	for _, strCIDR := range strCIDRs {
+		cidrSet, err := newCIDRSet(isV6, strCIDR, maskSize)
+		if err != nil {
+			return nil, err
+		}
+		cidrAllocators = append(cidrAllocators, cidrSet)
+	}
+	return cidrAllocators, nil
```

Comment:
This function must validate that `cidrAllocators` are not overlapping as otherwise, the behavior will be undefined.